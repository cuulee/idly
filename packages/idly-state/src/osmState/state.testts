import * as fs from 'fs';
import { deepFreeze } from '../misc/deepFreeze';
import { Entity, Node, Way } from '../osm/structures';
import { createNode, createWay } from './validate/createEntity';
import { setCreate } from './helper';
import { genIndex, logCreate } from './log/index';

const smallDataset = (freeze: boolean = false): Entity[] =>
  fs
    .readFileSync('./src/state2/mosm.jsonl', 'utf-8')
    .split('\n')
    .map(r => JSON.parse(r) as any)
    .map(r => (freeze ? deepFreeze(r) : r));

const biglDataset = (freeze: boolean = false): Entity[] =>
  fs
    .readFileSync('./src/state2/osm.jsonl', 'utf-8')
    .split('\n')
    .map(r => JSON.parse(r) as any)
    .map(r => (freeze ? deepFreeze(r) : r));

describe('addVirginEntities', () => {
  it('should add entities', () => {
    const table = entityTableCreate();
    entityTableAddVirginEntities(table, smallDataset(), '2');
    expect(table).toMatchSnapshot();
  });
  it('should add entities', () => {
    const data = biglDataset();
    const state = State.create();
    entityTableAddVirginEntities(state.entityTable, data.slice(0, 1e5), '2');
    getEntitiesInQuadkey(state, '2');
  });
});

describe('removeVirginEntities', () => {
  it('should remove', () => {
    const table = entityTableCreate();
    entityTableAddVirginEntities(table, smallDataset(), '2');
    const toRemove = ['n4133782719', 'n4133782714'];
    expect(
      [...entityTableRemoveVirginIndexes(table, toRemove).values()].map(
        r => r.id
      )
    ).toEqual(
      smallDataset()
        .map(r => r.id)
        .filter(r => toRemove.indexOf(r) === -1)
    );
  });
});

describe('getEntitiesInQuadkey', () => {
  it('should get get quadkey data', () => {
    const state = State.create();
    const data = smallDataset();
    const count = 6;
    entityTableAddVirginEntities(
      state.entityTable,
      data.slice(0, count),
      '32220'
    );
    entityTableAddVirginEntities(state.entityTable, data.slice(count), '12020');
    expect(getEntitiesInQuadkey(state, '')).toHaveLength(data.length);
    expect(getEntitiesInQuadkey(state, '3')).toHaveLength(count);
    expect(getEntitiesInQuadkey(state, '1')).toHaveLength(data.length - count);
  });
});

describe('defaultChecks', () => {
  const setup = () => {
    const table = entityTableCreate();
    const log = logCreate();
    const data = smallDataset();
    entityTableAddVirginEntities(table, data, '32220');
    return { log, table, data, state: State.create(log, table) };
  };

  it('it throws when shape is malformed', () => {
    const { log, table, state } = setup();

    const check = (entities: any[]) =>
      defaultChecks(state, entities).cata(err => {
        throw err;
      }, es => es);

    const node = table.get('n4133782723') as Node;
    expect(() =>
      check([
        {
          ...node,
          id: genIndex('n-4')(log),
          loc: null,
        } as any,
      ])
    ).toThrowErrorMatchingSnapshot();
    expect(() =>
      check([
        {
          id: 'w-4#0',
        } as any,
      ])
    ).toThrowErrorMatchingSnapshot();
  });
  it('it throws when virgin', () => {
    const { table, state } = setup();
    const check = (entities: any[]) =>
      defaultChecks(state, entities).cata(err => {
        throw err;
      }, es => es);
    const node = table.get('n4133782723') as Node;
    expect(() =>
      check([
        {
          ...node,
          id: 'n-5',
        } as any,
      ])
    ).toThrowErrorMatchingSnapshot();
  });
  it('it throws when no parent', () => {
    const { table, state } = setup();
    const check = (entities: any[]) =>
      defaultChecks(state, entities).cata(err => {
        throw err;
      }, es => es);
    const node = table.get('n4133782723') as Node;
    expect(() =>
      check([
        {
          ...node,
          id: 'n-5#0',
        } as any,
      ])
    ).toThrowErrorMatchingSnapshot();
  });
});

describe('commitModifiedEntities', () => {
  const setup = (quadkey = '32220') => {
    const state = State.create();
    const data = smallDataset();
    entityTableAddVirginEntities(state.entityTable, data, quadkey);
    return { table: state.entityTable, log: state.log, state, data };
  };

  it('works with empty commit', () => {
    const { log, table } = setup();
    const commit = commitModifiedEntities(table);
    expect(commit(log, [])).toEqual([setCreate()]);
  });

  describe('it adds a modified node', () => {
    const quadkey = '230';
    const { log, table, state } = setup(quadkey);
    const commit = commitModifiedEntities(table);
    const node = table.get('n4133782723') as Node;

    const newNode = createNode({
      ...node,
      id: genIndex(node.id)(log),
      loc: {
        lat: 23,
        lon: 35,
      },
    });

    expect(newNode).toMatchSnapshot();

    const logv0 = commit(log, [newNode]);
    expect(logv0).toMatchSnapshot();

    it('shouldnt modify the table', () => {
      expect(table).toEqual(setup().table.set(newNode.id, newNode));
    });

    it('new node is searchable with al; quadkeys', () => {
      expect(
        getEntitiesInQuadkey(state, quadkey).find(r => r.id === newNode.id)
      ).toEqual(newNode);
    });

    it('new node showsup with other quadkey', () => {
      expect(getEntitiesInQuadkey(state, '1').map(r => r.id)).toEqual([
        newNode.id,
      ]);
    });

    it('old node doesnt show up in any case', () => {
      expect(
        getEntitiesInQuadkey(state, quadkey).find(r => r.id === node.id)
      ).toEqual(undefined);
      expect(
        getEntitiesInQuadkey(state, '').find(r => r.id === node.id)
      ).toEqual(undefined);
    });
  });

  describe('multiple same commits shouldnt change  table ', () => {
    const { log, table } = setup();
    const commit = commitModifiedEntities(table);
    const way = table.get('w411770875') as Way;
    const newWay = createWay({
      ...way,
      id: genIndex(way.id)(log),
      tags: {
        newTag: 'newValue',
      },
    });
    const logv0 = commit(log, [newWay]);
    const tv0 = JSON.stringify(table);
    it('should not throw error if log remains same', () => {
      expect(commit(log, [newWay])).toEqual(logv0);
    });
    it('table should remain unchanged', () => {
      expect(tv0 === JSON.stringify(table)).toBe(true);
    });
    it('should throw if the new log is fed with same entity', () => {
      expect(() => commit(logv0, [newWay])).toThrowError();
    });
  });

  it('doesnt modifies the table when wrong data', () => {
    const { log, table } = setup();
    const commit = commitModifiedEntities(table);
    const node = table.get('n4133782723') as Node;
    expect(() =>
      commit(log, [
        {
          ...node,
          id: 'n-5#0',
        } as any,
      ])
    ).toThrowErrorMatchingSnapshot();
    expect(table).toEqual(setup().table);
  });

  describe('multiple related modifications', () => {
    const { log, table } = setup();
    const commit = commitModifiedEntities(table);
    const n4133782723 = table.get('n4133782723') as Node;

    const n4133782723vv0 = {
      ...n4133782723,
      id: genIndex(n4133782723.id)(log),
      tags: { ...n4133782723.tags, t: 'skdjaslkjdn' },
    };

    const n12 = table.get('n12') as Node;
    const n12vv0 = {
      ...n12,
      id: genIndex(n12.id)(log),
      tags: { ...n12.tags, t: 's232edn2' },
    };

    expect(commit(log, [n4133782723vv0, n12vv0])).toEqual([
      setCreate([n4133782723vv0.id, n12vv0.id]),
    ]);
    expect(table).toEqual(
      setup()
        .table.set(n12vv0.id, n12vv0)
        .set(n4133782723vv0.id, n4133782723vv0)
    );
  });
  it('throws error when one of multiple related modifications is malformed', () => {
    const { log, table } = setup();
    const commit = commitModifiedEntities(table);
    const n4133782723 = table.get('n4133782723') as Node;
    const n12 = table.get('n12') as Node;
    expect(() =>
      commit(log, [
        {
          ...n4133782723,
          id: 'n31',
          tags: { ...n4133782723.tags, t: 'skdjaslkjdn' },
        },
        {
          ...n12,
          id: genIndex(n12.id)(log),
          tags: { ...n12.tags, t: 's232edn2' },
        } as any,
      ])
    ).toThrowErrorMatchingSnapshot();
  });

  describe('series of modification ', () => {
    const quadkey = '12';
    const { log, table, state } = setup(quadkey);
    const commit = commitModifiedEntities(table);
    const n12 = table.get('n12') as Node;
    const n12vv0 = createNode({
      ...n12,
      id: genIndex(n12.id)(log),
      tags: {
        newTag: 'v0',
      },
    });
    const logv0 = commit(log, [n12vv0]);

    it('first version', () => {
      expect(logv0).toEqual([setCreate([n12vv0.id])]);
    });

    const n12vv1 = createNode({
      ...n12,
      id: genIndex(n12vv0.id)(logv0),
      tags: {
        newTag: 'v1',
      },
    });
    const logv1 = commit(logv0, [n12vv1]);

    it('n12 v1', () => {
      expect(logv1).toEqual([setCreate([n12vv1.id]), setCreate([n12vv0.id])]);
    });

    it('doesnt let add older version', () => {
      expect(() => commit(logv1, [n12vv1])).toThrowErrorMatchingSnapshot();
      expect(() => commit(logv1, [n12vv0])).toThrowErrorMatchingSnapshot();
      expect(() => commit(logv0, [n12])).toThrowErrorMatchingSnapshot();
    });

    const n4133782723 = table.get('n4133782723') as Node;
    const n4133782723vv0 = {
      ...n4133782723,
      id: genIndex(n4133782723.id)(logv1),
      tags: { ...n4133782723.tags, t: 'v0' },
    };
    const n12vv2 = createNode({
      ...n12,
      id: genIndex(n12.id)(logv1),
    });

    const logv2 = commit(logv1, [n12vv2, n4133782723vv0]);

    it('12 v2 and n4133782723 v0', () => {
      expect(logv2).toEqual([
        setCreate([n12vv2.id, n4133782723vv0.id]),
        setCreate([n12vv1.id]),
        setCreate([n12vv0.id]),
      ]);
    });

    it('quadkey search should show modified up', () => {
      const searchResult = getEntitiesInQuadkey(state, quadkey).map(r => r.id);
      expect(searchResult).toContain(n4133782723vv0.id);
      expect(searchResult).toContain(n12vv2.id);
      expect(searchResult).not.toContain(n4133782723.id);
      expect(searchResult).not.toContain(n12vv1.id);
      expect(searchResult).not.toContain(n12vv0.id);
      expect(searchResult).not.toContain(n12.id);
    });

    it('quadkey search for different area should show modified entities', () => {
      const searchResult = getEntitiesInQuadkey(state, '3').map(r => r.id);
      expect(searchResult).toContain(n4133782723vv0.id);
      expect(searchResult).toContain(n12vv2.id);
      expect(searchResult).toHaveLength(2);
    });
  });
});

describe('addVirginEntities', () => {
  const setup = () => {
    const table = entityTableCreate();
    let log = logCreate();
    const data = smallDataset();
    entityTableAddVirginEntities(table, data, '12');

    const commit = commitModifiedEntities(table);

    const n4133782723 = table.get('n4133782723') as Node;
    const n12 = table.get('n12') as Node;

    const n4133782723vv0 = {
      ...n4133782723,
      id: genIndex(n4133782723.id)(log),
      tags: { ...n4133782723.tags, tag: 'n4133782723vv0' },
    };
    const n12vv0 = {
      ...n12,
      id: genIndex(n12.id)(log),
      tags: { ...n12.tags, t: 's232edn2' },
    };
    log = commit(log, [n4133782723vv0, n12vv0]);

    const way = table.get('w411770875') as Way;

    const newWay = createWay({
      ...way,
      id: genIndex(way.id)(log),
      tags: {
        newTag: 'newValue',
      },
    });

    log = commit(log, [newWay]);
    const n12vv1 = {
      ...n12,
      id: genIndex(n12.id)(log),
      tags: { ...n12.tags, t: 'n12vv1' },
    };

    log = commit(log, [n12vv1]);

    return { log, table, state: State.create(log, table) };
  };
  setup();
  it('adding virgin entities should work', () => {
    //
  });
});
