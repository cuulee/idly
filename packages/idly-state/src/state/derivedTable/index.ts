import { Entity, EntityType } from 'idly-common/lib/osm/structures';

import { setCreate, setEqual } from '../../dataStructures/set';

export type DerivedTable = Map<string, Derived>;

export interface Derived {
  readonly entity: Entity;
  readonly parentWays: Set<string>;
  readonly parentRelations: Set<string>;
}

export type DerivedCreateType = (entity: Entity) => Derived;
export const derivedCreate: DerivedCreateType = entity => ({
  entity,
  parentRelations: setCreate<string>(),
  parentWays: setCreate<string>(),
});

// 1. Add expanded nodes to ways, this will remove dependency on table
//     note than nodes of ways is a derived data for ways and not table
//     in this file we derive everything and then compare, deriving nodes
///     everytime wont be a good idea

export type DerivedTableUpdateType = (
  entityTable: Map<string, Entity>,
  derivedTable?: Map<string, Derived>
) => DerivedTable;

export const derivedTableUpdate: DerivedTableUpdateType = (
  entityTable,
  derivedTable = new Map()
) => {
  const newTable = derive(entityTable);

  newTable.forEach((newVal, id) => {
    const prevVal = derivedTable.get(id);

    return isEqual(newVal, prevVal)
      ? newTable.set(id, prevVal as Derived) // puts the prevVal in newTable to preserve the instance for === caching
      : derivedTable.set(id, newVal);
  });

  return newTable;
};

export type DeriveParentsType = (
  entityTable: Map<string, Entity>
) => DerivedTable;
export const derive: DeriveParentsType = entityTable => {
  const derivedTable: Map<string, Derived> = new Map();

  entityTable.forEach(entity =>
    derivedTable.set(entity.id, derivedCreate(entity))
  );

  derivedTable.forEach(compute => {
    if (compute.entity.type === EntityType.WAY) {
      compute.entity.nodes.forEach(nodeId => {
        const val = derivedTable.get(nodeId);
        if (!val) {
          throw new Error(
            `${compute.entity.id} doesnt have node=${nodeId} in derivedTable`
          );
        }
        val.parentWays.add(compute.entity.id);
      });
    } else if (compute.entity.type === EntityType.RELATION) {
      compute.entity.members.forEach(({ id }) => {
        const entityComp = derivedTable.get(id);
        if (entityComp) {
          entityComp.parentRelations.add(compute.entity.id);
        }
      });
    }
  });
  return derivedTable;
};

export const isEqual = (a: Derived, b?: Derived) => {
  if (!a || !b) {
    return false;
  }
  if (a.entity !== b.entity) {
    return false;
  }
  if (a.entity.type === EntityType.NODE) {
    if (!setEqual(a.parentWays, b.parentWays)) {
      return false;
    }
  }
  if (!setEqual(a.parentRelations, b.parentRelations)) {
    return false;
  }
  return true;
};
