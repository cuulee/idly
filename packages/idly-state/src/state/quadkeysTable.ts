import { iterableFlattenToSet, mapFilter, setCreate } from '../helper';
import {
  OneToManyTable,
  oneToManyTableCreate,
  oneToManyTableFilter,
  oneToManyTableUpdateIndex,
  tableFindKey,
  tableForEach,
  tableRemove,
} from '../table';
import { ReadonlyOneToManyTable } from '../table/oneToMany';
import { tableAdd, tableValues } from '../table/regular';

export type QuadkeysTable = OneToManyTable<string>;
export type ReadonlyQuadkeysTable = ReadonlyOneToManyTable<string>;

/**
 * All the modifiedIds are stored in the
 * global root '' quadkey, the virgin Ids are always
 * stored in whatever quadkey they came from. The
 * advantage being modifiedIds being available for render
 * at all times.
 */

export function quadkeysTableAdd(
  t: QuadkeysTable,
  ids: string[],
  quadkey: string
) {
  if (ancestorProperFind(t, quadkey)) {
    return;
  }

  // Note: since '' quadkey is not a valid quadkey,
  // but it is a subset of every string. It becomes
  // a good place to search for everything.
  // we let it be used as key store and only append
  // to it.
  if (quadkey === '') {
    oneToManyTableUpdateIndex(t, quadkey, ids);
    return;
  }

  // TOFIX we wanna remove this
  // removes any descendant in favour of the parent quadkey
  removeAllDescendants(t, quadkey);

  tableAdd(setCreate(ids), quadkey, t);
}

export function removeAllDescendants(t: QuadkeysTable, quadkey: string) {
  tableForEach((_, id) => {
    if (isQuadkeyDescendant(quadkey, id)) {
      tableRemove(id, t);
    }
  }, t);
}

export const quadkeysTableCreate = (
  q = oneToManyTableCreate<string>()
): QuadkeysTable => q;

export const quadkeysTableCreateFrom = (t: QuadkeysTable, quadkeys: [string]) =>
  mapFilter((_, k) => quadkeys.indexOf(k) > -1, t);

export const quadkeysTableFlatten = (t: QuadkeysTable) =>
  iterableFlattenToSet<string>(tableValues(t));

export const quadkeysTableFindRelated = (
  t: QuadkeysTable,
  quadkeys: string[]
): ReadonlySet<string> =>
  quadkeys.some(q => q === '')
    ? setCreate()
    : iterableFlattenToSet(
        oneToManyTableFilter(
          (_, k) => k !== '' && quadkeys.some(q => isQuadkeyRelated(q, k)),
          t
        )
      );

export const isQuadkeyRelated = (q: string, maybeRelated: string) =>
  q === maybeRelated ||
  isQuadkeyDescendant(q, maybeRelated) ||
  isQuadkeyAncestor(q, maybeRelated);

export const isQuadkeyDescendant = (q: string, maybeDescendant: string) =>
  maybeDescendant.startsWith(q) && q !== maybeDescendant;

export const isQuadkeyAncestor = (q: string, maybeAncestor: string) =>
  q.startsWith(maybeAncestor) && q !== maybeAncestor;

export const isQuadkeyParent = (q: string, maybeParent: string) =>
  maybeParent === q.slice(0, -1);

export const isQuadkeyChild = (q: string, maybeChild: string) =>
  q === maybeChild.slice(0, -1);

// excludes '' as an ancestor
export const ancestorProperFind = (t: QuadkeysTable, quadkey: string) =>
  tableFindKey((_, id) => isQuadkeyAncestor(quadkey, id) && id !== '', t);
