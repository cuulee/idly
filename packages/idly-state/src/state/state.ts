import { Table } from '../table';
import { tableAdd, tableHas } from '../table/regular';

import {
  QuadkeysTable,
  quadkeysTableAdd,
  quadkeysTableCreate,
  quadkeysTableFindRelated,
} from './quadkeysTable';

export class State<T, M = any> {
  public static create<T, M = T>(
    elementTable: Table<T> = new Map(),
    quadkeysTable = quadkeysTableCreate(),
    metaTable: Table<M> = new Map()
  ) {
    return new State<T, M>(elementTable, quadkeysTable, metaTable);
  }

  // tslint:disable-next-line:variable-name
  private _elementTable: Table<T>;
  // tslint:disable-next-line:variable-name
  private _quadkeysTable: QuadkeysTable;
  // tslint:disable-next-line:variable-name
  private _metaTable = new Map<string, M>();

  private constructor(
    elementTable: Table<T>,
    quadkeysTable: QuadkeysTable,
    metaTable: Table<M>
  ) {
    this._elementTable = elementTable;
    this._quadkeysTable = quadkeysTable;
    this._metaTable = metaTable;
  }
  public getElementTable() {
    return this._elementTable;
  }
  public getMetaTable() {
    return this._metaTable;
  }
  public getQuadkeysTable() {
    return this._quadkeysTable;
  }

  public getElement(id: string) {
    return this._elementTable.get(id);
  }

  public getIdsByQuadkey(quadkey: string) {
    return this._quadkeysTable.get(quadkey);
  }

  public has(quadkey: string) {
    return this._quadkeysTable.has(quadkey);
  }

  // a single element can be on multiple quadkeys, but
  // changing an existing element is disallowed
  public add(getId: (t: T) => string, elements: T[], quadkey: string) {
    elements.forEach(
      e =>
        tableHas(getId(e), this._elementTable) ||
        tableAdd(e, getId(e), this._elementTable)
    );
    quadkeysTableAdd(this._quadkeysTable, elements.map(getId), quadkey);
  }

  public getQuadkey(quadkey: string) {
    return this._quadkeysTable.get(quadkey);
  }

  public getVisible(quadkeys: string[]) {
    return quadkeysTableFindRelated(this._quadkeysTable, quadkeys);
  }
}
