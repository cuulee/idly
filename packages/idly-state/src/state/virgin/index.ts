import { Table, tableBulkAdd } from '../../dataStructures/table';

import { iterableFlattenToSet } from '../../dataStructures/set';
import {
  QuadkeysTable,
  quadkeysTableAdd,
  quadkeysTableCreate,
  quadkeysTableGet,
} from './quadkeysTable';

export interface VirginState<T extends { id: string }> {
  readonly quadkeysTable: QuadkeysTable;
  readonly elements: Table<T>;
}

export function virginStateCreate<T extends { id: string }>(
  elements: Table<T> = new Map(),
  quadkeysTable = quadkeysTableCreate()
): VirginState<T> {
  return {
    elements,
    quadkeysTable,
  };
}

export function quadkeyGet(quadkeysTable: QuadkeysTable, quadkey: string) {
  return quadkeysTable.get(quadkey);
}

export function virginAddElements<T extends { id: string }>(
  elements: T[],
  quadkey: string,
  state: VirginState<T>
) {
  if (state.quadkeysTable.has(quadkey)) {
    return;
  }
  tableBulkAdd(elements.map((r: T): [string, T] => [r.id, r]), state.elements);
  quadkeysTableAdd(state.quadkeysTable, elements.map(r => r.id), quadkey);
}

export function virginGetInQuadkeys(
  quadkeysTable: QuadkeysTable,
  quadkeys: string[]
): ReadonlySet<string> {
  return iterableFlattenToSet(
    quadkeys
      .map(q => quadkeysTableGet(quadkeysTable, q))
      .filter((r): r is ReadonlySet<string> => !!r)
  );
}
