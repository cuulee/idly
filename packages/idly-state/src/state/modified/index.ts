import { Entity } from 'idly-common/lib/osm/structures';
import { baseId } from '../../dataStructures/log';
import {
  ReadonlyTable,
  Table,
  tableAdd,
} from '../../dataStructures/table/regular';
import { OsmState } from '../../type';
import { entitiesUnionBaseAndNodesOfWays } from '../virgin/helpers';

// http://localhost:8080/#18.73/40.7269/-74.00153 this just loads king st
// export type VirginTable = ReadonlyTable<Entity>;

// tofix make sure you only use readonly and readonly safe operators

// export function addVirgins(
//   state: OsmState,
//   entities: Entity[],
//   quadkey: string
// ): OsmState {
// if (entities.some(r => isNotVirgin(r.id))) {
//   throw new Error('only virgin entities can be added');
// }
// const isParentOfModified: Array<[Way, string[]]> = [];
// for (const e of entities) {
//   if (e.type === EntityType.NODE || state.modified.has(e.id)) {
//     continue;
//   }
//   if (e.type === EntityType.WAY) {
//     const foundNodes = e.nodes.filter(n => state.modified.has(n));
//     if (foundNodes.length > 0) {
//       isParentOfModified.push([e, foundNodes]);
//     }
//   }
// }
// const getBaseIds = e => [...e].map(r => modifiedIdGetBaseId(r));
// let log = state.log;
// let modified = state.modified;
// if (isParentOfModified.length > 0) {
//   modified = new Map(modified);
//   isParentOfModified.forEach(([entity, foundNodes]) => {
//     log = logRewrite(log, entity.id, new Set(foundNodes));
//     log
//       .slice(0)
//       .reverse()
//       .map(e => [...e])
//       .map(ids => [
//         ids[ids.map(modifiedIdGetBaseId).indexOf(entity.id)],
//         ids.filter(i => entity.nodes.indexOf(modifiedIdGetBaseId(i)) > -1),
//       ])
//       .filter(([id]) => id)
//       .forEach(([id, modifiedNodeIds]) => {
//         const newWay = {
//           ...entity,
//           id,
//           nodes: entity.nodes.map(r => {
//             if (modifiedNodeIds.find(mN => modifiedIdGetBaseId(mN) === r)) {
//               return modifiedNodeIds.find(
//                 mN => modifiedIdGetBaseId(mN) === r
//               );
//             }
//             return r;
//           }),
//         };
//         modified.set(newWay.id, newWay);
//         if (!modified.has(entity.id)) {
//           modified.set(entity.id, entity);
//         }
//       });
//   });
// }
// }

export function modifiedGetEntity(
  modified: ReadonlyTable<Entity>,
  id: string
): Entity | undefined {
  return modified.get(id);
}

function debugModifiedAddModifiedEntities(
  table: OsmState['modified'],
  virginTable: OsmState['virgin'],
  modifiedEntities: Entity[]
) {
  modifiedEntities.forEach(e => {
    const id = baseId(e.id);
    if (!table.has(id)) {
      if (!virginTable.elements.get(id)) {
        throw new Error('virgin table doesnt have ' + id);
      }
    }
  });
}

export function modifiedAddModifiedEntitiesAndRelated(
  modified: OsmState['modified'],
  virgin: OsmState['virgin'],
  modifiedEntities: Entity[]
): Table<Entity> {
  debugModifiedAddModifiedEntities(modified, virgin, modifiedEntities);

  return entitiesUnionBaseAndNodesOfWays(modifiedEntities, virgin)
    .filter(r => !modified.has(r.id))
    .reduce((cTable, cur) => tableAdd(cur, cur.id, cTable), new Map(modified));
}
