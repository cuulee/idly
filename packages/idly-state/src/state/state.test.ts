import {
  nodeFactory,
  wayFactory,
} from 'idly-common/lib/osm/entityFactory/index';
import { Entity } from 'idly-common/lib/osm/structures';
import { setCreate } from '../helper';
import { State } from './state';

const mapFromObj = <T>(o: any = {}): Map<string, T> =>
  Object.keys(o).reduce((prev, k) => {
    prev.set(k, o[k]);
    return prev;
  }, new Map());

const n1 = nodeFactory({
  id: 'n1',
});
const n2 = nodeFactory({
  id: 'n2',
});
const n3 = nodeFactory({
  id: 'n3',
});
const n4 = nodeFactory({
  id: 'n4',
});

const w1 = wayFactory({
  id: 'w1',
  nodes: ['n1', 'n2'],
});

const getId = (r: any) => r.id;
describe('constructing state and basic tests', () => {
  it('should create', () => {
    expect(State.create()).toEqual({
      _elementTable: mapFromObj(),
      _metaTable: mapFromObj(),
      _quadkeysTable: mapFromObj(),
    });
  });
  it('should create with given values', () => {
    const state = State.create<Entity>();
    state.add(getId, [n1, n2, w1], '123');
    expect(state).toMatchSnapshot();
  });
});

describe('getVisible', () => {
  describe('should get ids', () => {
    const state = State.create<Entity>();
    state.add(getId, [n1, n2, w1], '123');

    state.add(getId, [n3], '121');
    state.add(getId, [n4], '130');

    const n1Hash0 = nodeFactory({
      id: 'n1#0',
    });
    const n2Hash0 = nodeFactory({
      id: 'n2#0',
    });
    const w1Hash0 = wayFactory({
      id: 'w1#0',
      nodes: [n1Hash0.id, n2Hash0.id],
    });

    state.add(getId, [n1Hash0, n2Hash0, w1Hash0], '');

    expect(state.getQuadkey('')).toEqual(
      setCreate([n2Hash0, w1Hash0, n1Hash0].map(e => e.id))
    );

    expect(state.getVisible(['121'])).toEqual(setCreate([n3].map(e => e.id)));

    expect(state.getVisible(['130'])).toEqual(setCreate([n4].map(e => e.id)));

    expect(state.getVisible(['1'])).toEqual(
      setCreate([n1, n2, n3, n4, w1].map(e => e.id))
    );

    it('repeating state.add shouldnt affect', () => {
      state.add(getId, [n1Hash0, n2Hash0, w1Hash0], '');
      expect(state.getQuadkey('')).toEqual(
        setCreate([n2Hash0, w1Hash0, n1Hash0].map(e => e.id))
      );
      state.add(getId, [n3], '121');
      expect(state.getVisible(['121'])).toEqual(setCreate([n3].map(e => e.id)));
      state.getVisible([]);
    });
  });

  describe('should handle ids belonging to multiple quadkeys', () => {
    const state = State.create<Entity>();
    state.add(t => t.id, [n1, n2], '123');
    state.add(t => t.id, [w1], '122');

    expect(state.getVisible(['123'])).toEqual(setCreate(['n1', 'n2']));
    expect(state.getVisible(['122'])).toEqual(setCreate(['w1']));
    it('should not insert quadkey whose parent already exists', () => {
      state.add(t => t.id, [n3], '1234');
      expect(state.getVisible(['1234'])).toEqual(setCreate(['n1', 'n2']));
      expect(state.getElement('n3')).toEqual(n3);
    });
  });
});

describe('add', () => {
  it('doesnt overwrite existing elements', () => {
    const state = State.create<Entity>();
    state.add(r => r.id, [n1, n2, w1], '123');

    const n1Other = nodeFactory({
      id: 'n1',
    });

    expect(n1Other).not.toBe(n1);

    state.add(r => r.id, [n1Other], '12');

    expect(state.getElement('n1')).toBe(n1);
  });
});

// describe('shred', () => {
//   it('shouldnt change existing state', () => {
//     const state = State.create<Entity>();
//     state.add(t => t.id, [n1, n2], '12301201');
//     state.add(t => t.id, [w1], '12101200');
//     state.add(t => t.id, [w1, w2], '110231121');
//     state.add(t => t.id, [w3, r2], '100231121');
//     // console.log(JSON.stringify(state));
//     state.shred('12');

//     expect(state).toEqual(
//       (s => {
//         s.add(t => t.id, [n1, n2], '12301201');
//         s.add(t => t.id, [w1], '12101200');
//         s.add(t => t.id, [w1, w2], '110231121');
//         s.add(t => t.id, [w3, r2], '100231121');
//         return s;
//       })(State.create<Entity>())
//     );
//     state.shred('');
//     expect(state).toEqual(
//       (s => {
//         s.add(t => t.id, [n1, n2], '12301201');
//         s.add(t => t.id, [w1], '12101200');
//         s.add(t => t.id, [w1, w2], '110231121');
//         s.add(t => t.id, [w3, r2], '100231121');
//         return s;
//       })(State.create<Entity>())
//     );

//     expect(state.shred('100')).toMatchSnapshot();
//   });
//   it('should loose weight', () => {
//     let state = State.create<Entity>();
//     state.add(t => t.id, [n1, n2], '12301201');
//     state.add(t => t.id, [w1], '12101200');
//     state.add(t => t.id, [n3], '');
//     state.add(t => t.id, [w1, w2], '110231121');
//     state.add(t => t.id, [w3, r2], '100231121');
//     state.add(t => t.id, [w3, r2], '100131121');
//     state.add(t => t.id, [n5, n6, r2], '131');

//     state = state.shred('100');

//     expect(state).toEqual(
//       (s => {
//         s.add(t => t.id, [w3, r2], '100231121');
//         s.add(t => t.id, [w3, r2], '100131121');
//         return s;
//       })(State.create<Entity>())
//     );

//     state.add(t => t.id, [w3, r2], '100131121');
//     state.add(t => t.id, [w1], '12101200');
//     state.add(t => t.id, [n5, n6, r2], '131');

//     expect(state.shred('1')).toEqual(state);
//   });
// });
