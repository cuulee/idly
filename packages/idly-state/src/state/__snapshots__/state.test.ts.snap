// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`constructing state and basic tests should create with given values 1`] = `
State {
  "_elementTable": Map {
    "n1" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "n2" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "w1" => Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Object {},
      "type": "way",
    },
  },
  "_metaTable": Map {},
  "_quadkeysTable": Map {
    "123" => Set {
      "n1",
      "n2",
      "w1",
    },
  },
}
`;
